autonav_ws/src/autonav_odometry/src/particlefilter.py

```diff
@@ -0,0 +1,102 @@
+### particlefilter.py from last year ###
+# TODO could all the for-each-particle math in this code be done ✨🚀faster🚀✨ with numpy than vanilla python?
+from autonav_msgs.msg import MotorFeedback, GPSFeedback, Position
+import numpy as np
+import math
+import random
+
+
+class Particle:
+    def __init__(self, x=0, y=0, theta=0, weight=1) -> None:
+        self.x = x
+        self.y = y
+        self.theta = theta
+        self.weight = weight
+
+
+class ParticleFilter:
+    def __init__(self, latitudeLength, longitudeLength) -> None:
+        self.num_particles = 750
+        self.gps_noise = 0.45
+        self.odom_noise = (0, 0, 0.1)
+        # TODO should these parameters be in a config file in autonav_odometry/resource?
+        self.init_particles()
+        # self.first_gps = None
+
+        self.latitude_length = latitudeLength
+        self.longitude_length = longitudeLength
+
+    def init_particles(self, seedHeading: float = 0.0, useSeedHeading: bool = False):
+        self.first_gps = None
+        if useSeedHeading: # particles generally point in an initial direction
+            self.particles = [Particle(0, 0, seedHeading + np.random.normal(0, 0.1)) for i in range(self.num_particles)]
+        else: # particles point literally everywhere
+            self.particles = [Particle(0, 0, i / self.num_particles * 2 * math.pi) for i in range(self.num_particles)]
+
+    def feedback(self, feedback: MotorFeedback) -> tuple[float, float, float]:
+        sum_x = 0
+        sum_y = 0
+        sum_theta_x = 0
+        sum_theta_y = 0
+        sum_weight = 0
+
+        for particle in self.particles:
+            # update particles
+            # MotorFeedback gives robot-relative data (x is forward, y is left (???)) so we convert it to track-relative with trig magic  
+            # TODO why are we scaling down delta_x and delta_theta again?
```

Comment:
that would be a Tony question but basically because magic number
we might not need it with the swerve drive (or we might need a different magic number), that remains to be seen